In [13]:
import math

# PROBLEM: Given two complex numbers --> (a + bi), (c + di)
# Return the result of the following operations:
    # (1) Addition 
        # (a + c) + (b+d)i
            # (a + c) -> Real Numbers
            # (b + d) -> Imaginary Numbers
    # (2) Subtraction
        # (a + bi) - (c+di)
            # (a-c) -> Real Numbers
            # (b-d) -> Imaginary Numbers
    # (3) Multiplication
        # (a + bi)(c + di) 
            # ac + adi + bci + bd(i**2) ----> i**2 = -1
            # ac + (ad + bc)i -bd
            # (ac-bd) + (ad + bc)i
            
    # (4) Division
        # (a +bi) / (c + di)
            # Multiply the numerator and denominator by the complex conjugate
            # Example Case:
                # Complex Part -> (c + di)
                # Complex Conjugate -> (c - di)
                
            # (a + bi)(c-di) / (c + di)(c -di)
                # (ac - adi + bci + bdi^2) / (c^2 -cdi + cdi -d^2*i^2)
                #                     i^2 = -1                    i^2 = -1
                # (ac + (-ad+bc)i - bd)    /  (c^2 - d^2(-1))
                # [(ac-bd) + (bc-ad)i]   /  (c^2   + d^2)
                    # Real: (ac-bd)
                    # Imaginary: (bc-ad)i
    # (5) Modulus
        # Given a complex number:
        # ---> z = a + bi
            #  |z| = √(a^2 + b^2)
                #  math.sqrt(a**2 + b**2)
                    # a -> Real Part
                    # b -> Imaginary Part
 
class Complex(object):
    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary
        
    def __add__(self, no):
        real = self.real + no.real
        imaginary = self.imaginary + no.imaginary
        return Complex(real, imaginary)
           
    def __sub__(self, no):
        real = self.real - no.real
        imaginary = self.imaginary - no.imaginary
        return Complex(real, imaginary)
        
    def __mul__(self, no):
        # (a + bi)(c + di) 
            # ac + adi + bci + bd(i**2) ----> i**2 = -1
                 # ac + (ad + bc)i -bd
                  # (ac-bd) + (ad + bc)i
        real = (self.real * no.real) - (self.imaginary * no.imaginary)
        imaginary = (self.real * no.imaginary) + (self.imaginary * no.real)
        return Complex(real, imaginary)       

    def __truediv__(self, no):
         # (a +bi) / (c + di)
            # Multiply the numerator and denominator by the complex conjugate
            # Example Case:
                # Complex Part -> (c + di)
                # Complex Conjugate -> (c - di)
                
            # (a + bi)(c-di) / (c + di)(c -di)
                # (ac - adi + bci + bdi^2) / (c^2 -cdi + cdi -d^2*i^2)
                #                     i^2 = -1                    i^2 = -1
                # (ac + (-ad+bc)i - bd)    /  (c^2 - d^2(-1))
                # [(ac-bd) + (bc-ad)i]   /  (c^2   + d^2)
                    # Real: (ac-bd)
                    # Imaginary: (bc-ad)i
        divisor = float(no.real**2 + no.imaginary**2)
        real = ((self.real * no.real) - (self.imaginary*no.imaginary)) / divisor
        imaginary = ((self.imaginary*no.real) - (self.real*no.imaginary)) / divisor
        return Complex(real, imaginary)
    
    def mod(self):
        # z = a + bi
        # |z| = square_root(a**2 + b**2)
                            # a -> Real Part
                            # b -> Imaginary Part
        a = self.real
        b = self.imaginary
        z = math.sqrt(a**2 + b**2)
        return Complex(z, 0)

    def __str__(self):
        if self.imaginary == 0:
            result = "%.2f+0.00i" % (self.real)
        elif self.real == 0:
            if self.imaginary >= 0:
                result = "0.00+%.2fi" % (self.imaginary)
            else:
                result = "0.00-%.2fi" % (abs(self.imaginary))
        elif self.imaginary > 0:
            result = "%.2f+%.2fi" % (self.real, self.imaginary)
        else:
            result = "%.2f-%.2fi" % (self.real, abs(self.imaginary))
        return result

if __name__ == '__main__':
    print("[+] Enter Complex Number (C): ") 
    c = map(float, input().split()) # HackerRank Input -->  2 1 
    print("[+] Enter Complex Number (D):")  # HackerRank Input -->  5 6
    d = map(float, input().split())
    x = Complex(*c)
    y = Complex(*d)
    print(*map(str, [x+y, x-y, x*y, x/y, x.mod(), y.mod()]), sep='\n')

[+] Enter Complex Number (C): 
[+] Enter Complex Number (D):
6.00+6.00i
-4.00-2.00i
-3.00+14.00i
-0.07+0.15i
2.24+0.00i
6.40+0.00i



# PROBLEM: Given two complex numbers --> (a + bi), (c + di)
# Return the result of the following operations:
    # (1) Addition 
        # (a + c) + (b+d)i
            # (a + c) -> Real Numbers
            # (b + d) -> Imaginary Numbers
    # (2) Subtraction
        # (a + bi) - (c+di)
            # (a-c) -> Real Numbers
            # (b-d) -> Imaginary Numbers
    # (3) Multiplication
        # (a + bi)(c + di) 
            # ac + adi + bci + bd(i**2) ----> i**2 = -1
            # ac + (ad + bc)i -bd
            # (ac-bd) + (ad + bc)i
            
    # (4) Division
        # (a +bi) / (c + di)
            # Multiply the numerator and denominator by the complex conjugate
            # Example Case:
                # Complex Part -> (c + di)
                # Complex Conjugate -> (c - di)
                
            # (a + bi)(c-di) / (c + di)(c -di)
                # (ac - adi + bci + bdi^2) / (c^2 -cdi + cdi -d^2*i^2)
                #                     i^2 = -1                    i^2 = -1
                # (ac + (-ad+bc)i - bd)    /  (c^2 - d^2(-1))
                # [(ac-bd) + (bc-ad)i]   /  (c^2   + d^2)
                    # Real: (ac-bd)
                    # Imaginary: (bc-ad)i
    # (5) Modulus
 

# ===== FORMAT =====
Round both parts of the complex number to two decimals
(1) Real Part (A)
(2) Complex Part (B)
    - A + Bi

# ===== EXAMPLE INPUTS =====
>>> 2 1
>>> 5 6

# ===== EXAMPLE OUTPUTS ===== 
>>> 7.00+7.00i
>>> -3.00-5.00i
>>> 4.00+17.00i
>>> 0.26-0.11i
>>> 2.24+0.00i
>>> 7.81+0.00i

